Major improvements and fixes for companion fitting

kammerje opened the pull request:
This is a big PR comprising of multiple improvements and fixes:

1. Get pixel area in steradian directly from FITS file header instead of computing it from pySIAF pixel scale. The value is slightly different and I'm assuming that the JWST pipeline used what is written to the FITS file header so this will make things more consistent.
2. Enable contrast curve calculation for MIRI 4QPM, MIRI LYOT is still not implemented!
3. Bugfixes when injecting and recovering companions into multiple datasets at once, especially if using high-pass filtering.
4. Increased size of automatic blurring to 2 times the theoretical value, based on testing with beta Pic data this is necessary to not obtain Fourier ripples.
5. Added background subtraction step in companion fitting routine. Before, it could happen that the KLIP-subtracted data is negative in the background if there is a bright source such as an extended disk in the images. This lead to very poor companion fitting results and wrong photometry. There is an option to activate a dedicated background removal before fitting the PSFs. Improvement for beta Pic data are from contrast within ~15% to contrast within ~3%.
6. It is now possible to not save the stage 2 files when injecting and recovering companion, to limit the number of disk space required.
7. Centering has been changed to (shape-1)/2 after a discussion with @AarynnCarter and consistently works with the injection and recovery module.

kammerje commented:
This plot shows the new background removal step and how it improves the measured contrast (injected contrast = 1e-4):
[JWST_NIRCAM_NRCA2_F182M_MASKRND_MASK210R_SUB640A210R-bgest_c1.pdf](https://github.com/kammerje/spaceKLIP/files/13647188/JWST_NIRCAM_NRCA2_F182M_MASKRND_MASK210R_SUB640A210R-bgest_c1.pdf)

mperrin commented:
For point 1 about pixel scale, we ought to be able to do more than assume what the pipeline is doing. Let’s ask the developers. 

The subtle part is that there is not a single unique correct pixel area; it varies slightly over the field of view due to optical distortions. So we should also double check which field point is being referred to. These are going to be tiny differences I expect but we should quantify.

kammerje commented:
> For point 1 about pixel scale, we ought to be able to do more than assume what the pipeline is doing. Let’s ask the developers.
> 
> The subtle part is that there is not a single unique correct pixel area; it varies slightly over the field of view due to optical distortions. So we should also double check which field point is being referred to. These are going to be tiny differences I expect but we should quantify.

Yes, please do so! The main interest for me is to reverse exactly what the JWST pipeline did when it converted Jy to Jy/sr.

mperrin commented:
> > For point 1 about pixel scale, we ought to be able to do more than assume what the pipeline is doing. Let’s ask the developers.
> > The subtle part is that there is not a single unique correct pixel area; it varies slightly over the field of view due to optical distortions. So we should also double check which field point is being referred to. These are going to be tiny differences I expect but we should quantify.
> 
> Yes, please do so! The main interest for me is to reverse exactly what the JWST pipeline did when it converted Jy to Jy/sr.

It turns out the pipeline is getting the average pixel area for each mode from the PHOTOM calibration files in CRDS. See docs here for that calibration step: https://jwst-pipeline.readthedocs.io/en/stable/jwst/photom/main.html
Specifically the subsection entitled “Pixel area data”

maxwellmb commented:
> As an IMPORTANT note, the following code needs to be added to pyklip.rdi.py in order to make RDI work with high-pass filtering (after line 56):
> 
> if isinstance(highpass, bool): if highpass: data = high_pass_filter_imgs(data, numthreads=None) else: # should be a number if isinstance(highpass, (float, int)): highpass = float(highpass) fourier_sigma_size = (data.shape[1]/(highpass)) / (2_np.sqrt(2_np.log(2))) data = high_pass_filter_imgs(data, numthreads=None, filtersize=fourier_sigma_size)

Yesterday I put in a pull request with a similar fix within pyklip!

kammerje commented:
@AarynnCarter @semaphoreP @mperrin (tagging all relevant people)

After some time of silence, I finally found the time to finish preparing this big PR from `dev/jk` into `develop`. In preparation for this PR, I have already merged `develop` into `dev/jk` a few days ago, resolved all the conflicts, tested things with the ERS data, and included a couple of additional fixes and improvements. Below, you can find a list of all the updates that this PR includes. Items in bold are new compared to the list posted at the top of this PR (several months ago). **I recommend merging this PR as soon as possible before `develop` diverts too far from `dev/jk` again and new merging conflicts arise!**

As an additional note for @semaphoreP, this PR is fully compatible with the `pyKLIP` `jwst` branch which I also updated this week. I reverted back to the previous version which does not need `pysiaf` as a dependency and also implemented a fallback option for when the SVO FPS is unavailable (see [here](https://bitbucket.org/pyKLIP/pyklip/src/jwst/)).

1. Get pixel area in steradian directly from FITS file header instead of computing it from `pysiaf` pixel scale and propagate it through the `spaceKLIP` database.
2. Enable contrast curve computation for MIRI 4QPM (simply copying @AarynnCarter's implementation), MIRI LYOT is still not implemented!
3. Bugfixes when injecting and recovering companions into multiple datasets at once, especially if using high-pass filtering.
4. Bugfixes when computing automatic blurring factor. As noted by @mperrin, the previous implementation was wrong. The automatic blurring factor is now set to 1.5 times the theoretically required value based on experience with the beta Pic data.
5. Add background subtraction step in companion fitting routine. Before, it could happen that the KLIP-subtracted data is negative in the background if there is a bright source such as an extended disk in the images. This lead to very poor companion fitting results and wrong photometry. There is now an option to activate a dedicated background removal step before fitting the PSFs.
6. Enabling possibility to not save stage 2 files when injecting and recovering companions to limit the required disk space.
7. Change centering to (shape - 1) / 2 after a discussion with @AarynnCarter. Consistently works with the injection and recovery module.
**8. Ensuring compatibility with the `pyKLIP` JWST class and enabling processing of data without `PIXAR_A2` SCI header keyword while not invoking additional dependencies on `pyKLIP`. This is required to process data for which the JWST pipeline photometry step has been skipped for instance.
9. Update installation instructions on GitHub (not on Read The Docs!).
10. Add fallback options for all SVO FPS calls for the case that the service is temporarily unavailable.
11. Pull PSF masks directly from CRDS instead of computing them manually in `make_psfmasks.py`.**

AarynnCarter commented on spaceKLIP/imagetools.py:
```diff
@@ -2071,6 +2121,13 @@ def create_rec_mask(h, w, center=None, z=None):
+                except:
+                    ref_image = pyfits.getdata(align_to_file, 0)
+                if ref_image.ndim == 3:
+                    ref_image = ref_image[0]
```
Do we want to just take one integration, or a median of them?

kammerje replied:
Changed to median.

AarynnCarter commented on spaceKLIP/starphot.py:
```diff
@@ -165,7 +166,16 @@ def get_stellar_magnitudes(starfile,
     # http://svo2.cab.inta-csic.es/theory/fps/
     filts = []
     zeros = []
-    filter_list = SvoFps.get_filter_list(facility='JWST', instrument=instrume)
```
I think the SVO changes are breaking our CI tests, could you take a look?

kammerje replied:
I have no clue about this CI stuff so this needs to be done by someone else please. :-)

AarynnCarter replied:
I'll take a look :) But, I added more comments for you!

AarynnCarter replied:
Okay this is fixed now!

AarynnCarter commented on spaceKLIP/analysistools.py:
```diff
@@ -1323,8 +1370,81 @@ def extract_companions(self,
+                            fma.noise_map[fma.noise_map == 0.] = noise_map_max
+                            
+                            # Run the MCMC fit.
+                            nwalkers = 50
```
On second thought, looks like this is already done further up. So this might be overwriting them.

kammerje replied:
Fixed.

kammerje commented on requirements.txt:
```diff
@@ -5,8 +5,8 @@ matplotlib>=3.7.1
 synphot
 stsynphot
-scipy
+scipy==1.13.1
```
That was because `pyKLIP` was not compatible with `SciPy` 1.14, but I think Jason fixed this so we can remove it.

kammerje commented on requirements.txt:
```diff
@@ -5,8 +5,8 @@ matplotlib>=3.7.1
 webbpsf
 webbpsf_ext>=1.2.0
-pyklip @ git+https://bitbucket.org/pyKLIP/pyklip.git@jwst
+pyklip@git+https://bitbucket.org/pyKLIP/pyklip.git@jwst
```
Yes, that should work now!

kammerje commented on spaceKLIP/analysistools.py:
```diff
@@ -841,6 +852,8 @@ def extract_companions(self,
                            fitkernel='diag',
                            subtract=True,
                            inject=False,
+                           remove_background=True,
```
Good point, I must have missed this. It should be `False` by default.

kammerje commented on spaceKLIP/analysistools.py:
```diff
@@ -1170,6 +1198,14 @@ def extract_companions(self,
+                        # sx = 3
+                        # sy = 3
+                        # offsetpsf = 1. / (2. * np.pi * sx * sy) * np.exp(-((xx - 32)**2. / (2. * sx**2.) + (yy - 32)**2. / (2. * sy**2.)))
+                        # offsetpsf *= peak / np.nanmax(offsetpsf)
```
Done!

kammerje commented on spaceKLIP/analysistools.py:
```diff
@@ -1267,8 +1304,18 @@ def extract_companions(self,
+                        # ramp = np.arange(65)
+                        # xx, yy = np.meshgrid(ramp, ramp)
+                        # sx = 3
+                        # sy = 3
```
Done!

kammerje commented on spaceKLIP/analysistools.py:
```diff
@@ -1323,8 +1370,81 @@ def extract_companions(self,
 
                     # Fit the FM PSF to the KLIP-subtracted data.
                     if inject == False:
+                        fitboxsize = 35  # pix
```
Fixed.

kammerje commented on spaceKLIP/analysistools.py:
```diff
@@ -1323,8 +1370,81 @@ def extract_companions(self,
+                            # Initialize pyKLIP FMAstrometry class.
+                            fma = fitpsf.FMAstrometry(guess_sep=guess_sep,
+                                                      guess_pa=guess_pa,
+                                                      fitboxsize=fitboxsize)
```
Fixed.

kammerje commented on spaceKLIP/analysistools.py:
```diff
@@ -1323,8 +1370,81 @@ def extract_companions(self,
+                            fma.generate_data_stamp(data=data_frame,
+                                                    data_center=[data_centx, data_centy],
+                                                    dr=dr,
+                                                    exclusion_radius=exclusion_radius)
```
Fixed.

kammerje commented on spaceKLIP/analysistools.py:
```diff
@@ -1323,8 +1370,81 @@ def extract_companions(self,
+                                                    data_center=[data_centx, data_centy],
+                                                    dr=dr,
+                                                    exclusion_radius=exclusion_radius)
+                            corr_len_label = r'$l$'
```
Not sure what you mean by this.

kammerje replied:
Still not sure about this.

kammerje commented on spaceKLIP/analysistools.py:
```diff
@@ -1323,8 +1370,81 @@ def extract_companions(self,
+                            # Estimate the background level from those pixels
+                            # in the KLIP-subtracted data which have a small
+                            # flux in the best fit FM PSF.
+                            hsz = 35
```
Done!

kammerje commented on spaceKLIP/analysistools.py:
```diff
@@ -1344,26 +1464,25 @@ def extract_companions(self,
                                 nwalkers = kwargs['nwalkers']
                             if 'nburn' not in kwargs.keys() or kwargs['nburn'] is None:
-                                nburn = 100
+                                nburn = 400
```
I'm always using 400 but happy to go to 100 if you say this is sufficient.

kammerje replied:
Already changed back to 100 in the code.

kammerje commented on spaceKLIP/analysistools.py:
```diff
@@ -1533,97 +1652,130 @@ def extract_companions(self,
+                            inputflux = con * np.array(all_offsetpsfs_nohpf)  # positive to inject companion
                             fileprefix = 'INJECTED-' + key
                         else:
                             ra = tab[-1]['RA']  # arcsec
```
Yes, because this leaves the flexibility to inject _any_ companion, not just the one that was fitted for.

kammerje commented on spaceKLIP/database.py:
```diff
@@ -481,10 +486,13 @@ def read_jwst_s012_data(self,
-                            maskfile = 'NONE'
+                        # maskpath = APERNAME[ww][j] + '_' + FILTER[ww][j] + '.fits'
+                        # maskfile = os.path.join(maskbase, maskpath)
+                        # if not os.path.exists(maskfile):
```
Fixed.

kammerje commented on spaceKLIP/database.py:
```diff
@@ -1021,8 +1036,10 @@ def print_obs(self,
+            print_tab['XOFFSET'] *= 1e3
             print_tab['XOFFSET'] = np.round(print_tab['XOFFSET'])
             print_tab['XOFFSET'][print_tab['XOFFSET'] == 0.] = 0.
+            print_tab['YOFFSET'] *= 1e3
```
This is just for printing, so it depends which unit you want to print. I would prefer mas since these shifts are typically rather small.

AarynnCarter commented on spaceKLIP/analysistools.py:
```diff
@@ -384,7 +391,10 @@ def raw_contrast(self,
+                        else:
+                            columns.append(cons_mask[i])
                         names.append(f'contrast, N_kl={klmode}')
                         columns.append(cons_mask[i])
```
Does this line need to be here? Looks like we are duplicating the if else statement from immediately above.

kammerje replied:
Hm, I don't see a similar if else statement before.

AarynnCarter replied:
I think I mean that you have the if else, then add cons_mask[i] on line 399 anyway

kammerje replied:
Ah, now I understand! What happens currently is that if a PSF mask file is available, both the raw contrast curve and the one taking into account the PSF mask file are saved. However, if not, then the same raw contrast curve is saved as both.

kammerje replied:
@AarynnCarter I think this might have been a merge conflict that I didn't resolve properly, or the automatic merging messed it up. It should be fixed now.

AarynnCarter commented on spaceKLIP/analysistools.py:
```diff
@@ -1004,9 +1031,9 @@ def extract_companions(self,
+                                          fov_pix=157,
                                           oversample=2,
-                                          sp=sed,
+                                          sp=None,
```
Not sure I understand this, why do we always want to spectral type set to None?

AarynnCarter replied:
This is extract_companions(), so I think it's making a model of an offaxis companion, which should use the planet SED.

kammerje replied:
See [this](https://github.com/kammerje/spaceKLIP/commit/701cd0f516b24a6c8e6676911356cad54f2f70bb) commit for a resolution of this bug.

AarynnCarter commented on spaceKLIP/analysistools.py:
```diff
@@ -1004,9 +1031,9 @@ def extract_companions(self,
                                           filt,
                                           date=date,
-                                          fov_pix=65,
+                                          fov_pix=157,
```
Also curious whether there is some reason behind using an FOV of 157 pixels.

kammerje replied:
Not sure where this change is coming from, I will revert it.

AarynnCarter commented on spaceKLIP/analysistools.py:
```diff
@@ -1289,42 +1321,121 @@ def extract_companions(self,
                         yrange = kwargs['yrange']
                     if 'frange' not in kwargs.keys() or kwargs['frange'] is None:
-                        frange = [-1e-2, 1e2] # * guess_flux
+                        frange = 10.  # mag
```
Can you explain why frange is now in units of magnitude?

kammerje replied:
Both inputs are allowed by `pyKLIP`. Do you prefer one of them?

AarynnCarter replied:
I'd say for now let stick to the default, unless you have a strong preference otherwise.

kammerje replied:
Can you confirm that `frange = [-1e-2, 1e2] # * guess_flux` makes sense? For me the default was always 10 mag. Why should it be allowed to be negative?

AarynnCarter replied:
I think in this case the frange is relative to the input. So two magnitudes higher or lower. My guess is 10 mag is the same thing?

AarynnCarter replied:
Maybe @semaphoreP can weigh in

semaphoreP replied:
The short answer is that I think the default should be `frange = 1` or `frange = 2`. Details below:

Hmm the docstring for this is a bit confusing, so that's something we should fix in pyklip. We don't use astronomy magnitudes, but it's also not `frange = [-1e-2, 1e2] # * guess_flux` . The code is literally:

```
if np.size(df) == 2:
            self.bounds.append(df)
        else:
            self.bounds.append([self.guess_flux / (10.**df), self.guess_flux * (10**df)])
```

So, either you pass in the explicit prior bound on the flux to scale the FM PSF to fit the data, or you indicate how many power of 10 the prior should vary from the best guess scale factor between the FM PSF and the data (computed in FitPSF as self.guess_flux).

AarynnCarter replied:
Thanks Jason! I've set it to 2 :)

AarynnCarter commented on spaceKLIP/imagetools.py:
```diff
@@ -1413,10 +1419,21 @@ def blur_frames(self,
+                            wave_min = self.database.obs[key]['CWAVEL'][j] - self.database.obs[key]['DWAVEL'][j]  # micron
+                            fwhm_current = wave_min * 1e-6 / diam * 180. / np.pi * 3600. / self.database.obs[key]['PIXSCALE'][j]  # pix
+                            fwhm_desired = 2.3  # pix; see, e.g., Pawley 2006
+                            fwhm_desired *= 1.5
```
Not sure I understand where this factor is coming from, could you expand in a comment?

kammerje replied:
Done.

AarynnCarter commented on spaceKLIP/pyklippipeline.py:
```diff
@@ -127,18 +130,19 @@ def run_obs(database,
+                    dataset = JWSTData(filepaths, psflib_filepaths, highpass=kwargs_temp['highpass'])
+                    kwargs_temp['dataset'] = dataset
+                    kwargs_temp['aligned_center'] = dataset._centers[0]
+                    kwargs_temp['psf_library'] = dataset.psflib
```
Not following why this was moved inside the loop. Doesn't that just slow things down? Or is there a reason?

semaphoreP replied:
pyklip is designed so you can run the same dataset through klip_dataset multiple times and it's fine.

AarynnCarter replied:
@kammerje I think that's what I meant, it looks like you've moved this inside the mode annuli subsections loop, so I think we should move it back outside to save time when calling JWSTData multiple times.

kammerje replied:
Thanks both, I fixed it.
